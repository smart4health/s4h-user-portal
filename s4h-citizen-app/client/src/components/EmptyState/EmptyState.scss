.EmptyState {
  display: flex;
  flex-direction: column;
  margin: auto;
  max-width: 304px;

  &__header {
    padding: var(--space-l) 0 calc(var(--space-xxl) + var(--space-xs)) 0;
  }

  &__content {
    padding-bottom: var(--space-l);
  }

  &__header + &__content,
  &__content + &__footer,
  &__header + &__footer {
    padding-top: 0;
  }

  &__footer {
    display: flex;
    flex-direction: column;
    align-items: stretch;
    justify-content: stretch;
    margin-top: 0;

    button {
      flex: 1 0 auto;
      margin: 0 4px 16px;

      &:first-child {
        margin-left: 0;
      }

      &:last-child {
        margin-right: 0;
      }

      @include tablet-up {
        &:only-child {
          flex: none;
          margin-left: auto;
        }
      }
    }
  }
}
