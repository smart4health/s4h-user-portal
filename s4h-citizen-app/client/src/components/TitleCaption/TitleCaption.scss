.TitleCaption {
  &__title {
    color: var(--color-primary);
    font-size: 21px;
    font-weight: 500;
    text-align: center;
    margin: 0;

    @include tablet-up {
      font-size: 24px;
    }
  }

  &__subtitle {
    color: var(--color-neutral);
    font: var(--typography-desktop-body-level-1);
    text-align: center;
    margin: 0;
  }

  &__title + &__subtitle {
    margin-top: 16px;
  }
}
