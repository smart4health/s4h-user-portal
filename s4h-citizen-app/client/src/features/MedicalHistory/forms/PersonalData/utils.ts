import { PersonalData } from '@d4l/s4h-fhir-xforms';
/* eslint-disable import/prefer-default-export */
import classNames from 'classnames';

export const valueSetUrls = {
  gender: 'http://hl7.org/fhir/ValueSet/administrative-gender',
  blood: 'http://loinc.org/vs/LL2419-1',
  rh: 'http://loinc.org/vs/LL360-9',
};

export type Option = {
  value: string;
  text: string;
};

const FEMALE = 'female';
const MALE = 'male';
const OTHER = 'other';
const UNKNOWN = 'unknown';

const BLOOD_0 = '0';
const BLOOD_A = 'A';
const BLOOD_B = 'B';
const BLOOD_AB = 'AB';

const BLOOD_RHESUS_FACTOR_POSITIVE = 'Positive';
const BLOOD_RHESUS_FACTOR_NEGATIVE = 'Negative';

const valueSetCodes: { [key: string]: string } = {
  'LA19708-9': BLOOD_0,
  'LA19709-7': BLOOD_B,
  'LA19710-5': BLOOD_A,
  'LA28449-9': BLOOD_AB,
  'LA6576-8': BLOOD_RHESUS_FACTOR_POSITIVE,
  'LA6577-6': BLOOD_RHESUS_FACTOR_NEGATIVE,
  female: FEMALE,
  male: MALE,
  other: OTHER,
  unknown: UNKNOWN,
};

export const mapValueSetCodeToTranslationKey = (code: string): string =>
  valueSetCodes[code];

export const fieldNames = {
  firstName: 'firstName',
  lastName: 'lastName',
  dateOfBirth: 'dateOfBirth',
  occupation: 'occupation',
  gender: 'gender',
  weight: 'weight',
  height: 'height',
  blood: 'blood',
  rh: 'rh',
};

export type FormValues = {
  firstName: string | undefined;
  lastName: string | undefined;
  dateOfBirth: string | undefined;
  occupation: string | undefined;
  gender: 'male' | 'female' | 'other' | 'unknown' | undefined;
  blood: 'LA19708-9' | 'LA19709-7' | 'LA19710-5' | 'LA28449-9' | undefined;
  rh: 'Positive' | 'Negative' | undefined;
  weight: string | undefined;
  height: string | undefined;
};

export type ValidationRules = {
  required?: boolean;
  min?: number;
  max?: number;
  maxlength?: number;
};

export const validationRules: { [key: string]: ValidationRules } = {
  [fieldNames.firstName]: {
    required: true,
  },
  [fieldNames.lastName]: {
    required: true,
  },
  [fieldNames.dateOfBirth]: {
    required: true,
  },
  [fieldNames.occupation]: {
    required: true,
  },
  [fieldNames.gender]: {
    required: true,
  },
  [fieldNames.weight]: {
    required: true,
    min: 1,
    max: 300,
  },
  [fieldNames.height]: {
    required: true,
    min: 1,
    max: 300,
  },
  [fieldNames.blood]: {
    required: true,
  },
  [fieldNames.rh]: {
    required: true,
  },
};

export const formToPersonalData = (formState: FormValues): PersonalData => {
  return {
    firstName: formState.firstName,
    lastName: formState.lastName,
    gender: formState.gender,
    dateOfBirth: formState.dateOfBirth,
    height: {
      value: parseInt(formState.height ? formState.height : '', 10),
      unit: 'cm',
    },
    weight: {
      value: parseInt(formState.weight ? formState.weight : '', 10),
      unit: 'kg',
    },
    bloodGroup: {
      system: 'http://loinc.org',
      code: formState.blood,
    },
    bloodRhesus: {
      system: 'http://loinc.org',
      code: formState.rh,
    },
    occupation: formState.occupation,
  };
};

export const personalDataToFormState = (personalData: PersonalData) => ({
  [fieldNames.firstName]: personalData?.firstName,
  [fieldNames.lastName]: personalData?.lastName,
  [fieldNames.dateOfBirth]: personalData?.dateOfBirth,
  [fieldNames.occupation]: personalData?.occupation,
  [fieldNames.gender]: personalData?.gender,
  [fieldNames.blood]: personalData?.bloodGroup?.code,
  [fieldNames.rh]: personalData?.bloodRhesus?.code,
  [fieldNames.weight]: personalData?.weight?.value,
  [fieldNames.height]: personalData?.height?.value,
});

export const getQuestionClass = (
  isReadOnly: boolean,
  hasBeenAnswered: boolean,
  additionalClasses = ''
) =>
  classNames(
    'MedicalHistoryQuestion',
    additionalClasses,
    {
      'MedicalHistoryQuestion--readonly': isReadOnly,
    },
    {
      'MedicalHistoryQuestion--answered': hasBeenAnswered,
    }
  );

export const formatDate = (year: string, month: string, day: string) =>
  `${year}${month ? `-${month}${day ? `-${day}` : ''}` : ''}`;
