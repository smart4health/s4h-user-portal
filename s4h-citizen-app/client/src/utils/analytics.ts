import ReactPiwik, { PiwikOptions } from 'react-piwik'; // '@types/react-piwik@latest' is not in the npm registry.
import config from '../config';
import settings, { SettingKeys } from './settings';

let matomo: any = null;

interface TrackingEvents {
  [key: string]: TrackingEvent;
}

type TrackingEvent = (string | number)[];

// The following mappings or names do not necessarily have to make sense.
// We're trying to map our events here to ones previously defined by mobile apps
// VAR_NAME_TERRA: ['Event category Matomo', 'Event action Matomo']
export const TRACKING_EVENTS: TrackingEvents = {
  DOCUMENT_UPLOAD_START: ['Documents - Save', 'Create document tapped'],
  DOCUMENT_UPLOAD_CANCEL: ['Documents - Save', 'Cancel create tapped'],
  DOCUMENT_UPLOAD_SUCCESS: ['Documents - Save', 'Document Save Success'],
  DOCUMENT_UPLOAD_ERROR: ['Documents - Save', 'Document Save Error'],
  DOCUMENT_EDIT_START: ['Documents - Editing', 'Edit document tapped'],
  DOCUMENT_EDIT_CANCEL: ['Documents - Editing', 'Cancel editing tapped'],
  DOCUMENT_EDIT_SUCCESS: ['Documents - Editing', 'Edit document success'],
  DOCUMENT_EDIT_ERROR: ['Documents - Editing', 'Edit document error'],
  DOCUMENT_DOWNLOAD_START: ['Documents - Download', 'Download document tapped'],
  DOCUMENT_DOWNLOAD_CANCEL: ['Documents - Download', 'Cancel download tapped'],
  DOCUMENT_DOWNLOAD_SUCCESS: ['Documents - Download', 'Download document success'],
  DOCUMENT_DOWNLOAD_ERROR: ['Documents - Download', 'Download document error'],
  DOCUMENT_ADD_START: ['Documents - Add', 'Add files tapped'],
  DOCUMENT_ADD_SUCCESS: ['Documents - Add', 'Add files success'],
  DOCUMENT_ADD_ERROR: ['Documents - Add', 'Add files error'],
  MEDICATION_DOWNLOAD_START: ['Medication - Download', 'Medication download tapped'],
  MEDICAL_HISTORY_PERSONAL_DATA_START: [
    'Medical History',
    'Edit personal_data tapped',
  ],
  MEDICAL_HISTORY_PERSONAL_DATA_ERROR: ['Medical History', 'personal_data error'],
  MEDICAL_HISTORY_PERSONAL_DATA_CANCEL: [
    'Medical History',
    'Cancel personal_data tapped',
  ],
  MEDICAL_HISTORY_PERSONAL_DATA_SUCCESS: [
    'Medical History',
    'Save personal_data tapped',
  ],
  MEDICAL_HISTORY_CONDITIONS_START: ['Medical History', 'Edit conditions tapped'],
  MEDICAL_HISTORY_CONDITIONS_ERROR: ['Medical History', 'conditions error'],
  MEDICAL_HISTORY_CONDITIONS_CANCEL: ['Medical History', 'Cancel conditions tapped'],
  MEDICAL_HISTORY_CONDITIONS_SUCCESS: ['Medical History', 'Save conditions tapped'],
  MEDICAL_HISTORY_RISK_FACTORS_START: [
    'Medical History',
    'Edit risk_factors tapped',
  ],
  MEDICAL_HISTORY_RISK_FACTORS_ERROR: ['Medical History', 'risk_factors error'],
  MEDICAL_HISTORY_RISK_FACTORS_CANCEL: [
    'Medical History',
    'Cancel risk_factors tapped',
  ],
  MEDICAL_HISTORY_RISK_FACTORS_SUCCESS: [
    'Medical History',
    'Save risk_factors tapped',
  ],
  MEDICAL_HISTORY_EMERGENCY_CONTACT_START: [
    'Medical History',
    'Edit emergency_contact tapped',
  ],
  MEDICAL_HISTORY_EMERGENCY_CONTACT_ERROR: [
    'Medical History',
    'emergency_contact error',
  ],
  MEDICAL_HISTORY_EMERGENCY_CONTACT_CANCEL: [
    'Medical History',
    'Cancel emergency_contact tapped',
  ],
  MEDICAL_HISTORY_EMERGENCY_CONTACT_SUCCESS: [
    'Medical History',
    'Save emergency_contact tapped',
  ],
  INITIATE_SHARING_START: ['Sharing', 'Start Sharing tapped'],
  INITIATE_SHARING_SUCCESS: ['Sharing', 'Start Sharing success'],
  SHARING_START: ['Sharing', 'Provide access tapped'],
  SHARING_SUCCESS: ['Sharing', 'Sharing success'],
  REVOKE_START: ['Sharing', 'Revoke tapped'],
  REVOKE_SUCCESS: ['Sharing', 'Revoke confirm tapped'],
  REVOKE_CANCEL: ['Sharing', 'Revoke cancel tapped'],
  NEW_SHARING_START: ['Sharing', 'Start new sharing session tapped'],
  OUTAPP_VIEWDATA_START: ['Sharing', 'View shared data tapped'],
  OUTAPP_VIEWDATA_SUCCESS: ['Sharing', 'View shared data success'],
  SHAREE_VIEW_DOWNLOAD_DATA: ['Sharee View', 'Download Data', 'Clicked'],
};

const isProduction: boolean = config.IS_PRODUCTION;
const isStaging: boolean = config.IS_STAGING;
const isLocalhost: boolean =
  config.ANALYTICS_ENABLE_LOCALHOST &&
  config.ANALYTICS_SITE_URL.startsWith('http://localhost:');

export const pushTrackingEvent = (trackingEvent: TrackingEvent) => {
  const isProduction = config.REACT_APP_ENVIRONMENT === 'production';

  if (!matomo || !ReactPiwik.push || !trackingEvent) {
    if (!isProduction) {
      console.log('TrackingEvent debug 🧐', trackingEvent);
    }
    return false;
  }
  ReactPiwik.push(['trackEvent', ...trackingEvent]);
  return true;
};
const initMatomo = () => {
  matomo = new ReactPiwik({
    url: config.ANALYTICS_SITE_URL,
    siteId: isProduction
      ? config.ANALYTICS_SITE_ID_PRODUCTION
      : isStaging
      ? config.ANALYTICS_SITE_ID_STAGING
      : config.ANALYTICS_SITE_ID_LOCALHOST,
    trackErrors: false,
    enableLinkTracking: true,
    jsFilename: 'matomo.js',
    phpFilename: 'matomo.php',
  } as PiwikOptions);

  ReactPiwik.push([
    'setCustomDimension',
    config.ANALYTICS_PRODUCTS_DIMENSION_KEY,
    config.ANALYTICS_PRODUCTS_DIMENSION_VALUE,
  ]);
  ReactPiwik.push([
    'setCustomDimension',
    config.ANALYTICS_PROJECTS_DIMENSION_KEY,
    config.ANALYTICS_PROJECTS_DIMENSION_VALUE,
  ]);
  ReactPiwik.push(['alwaysUseSendBeacon']);
  ReactPiwik.push(['enableCrossDomainLinking']);
  isProduction &&
    ReactPiwik.push(['setDomains', config.ANALYTICS_DOMAINS_PRODUCTION]);
  isStaging && ReactPiwik.push(['setDomains', config.ANALYTICS_DOMAINS_STAGING]);
  ReactPiwik.push(['trackPageView']);
  ReactPiwik.push(['enableHeartBeatTimer']);
};

const connectAnalytics = (browserHistory: any) => {
  if (!settings.acceptsTracking || (!isProduction && !isStaging && !isLocalhost)) {
    matomo && matomo.disconnectFromHistory();
    return false;
  }

  !matomo && initMatomo();
  matomo.connectToHistory(browserHistory);
  return true;
};

const setupAnalytics = (browserHistory: any) => {
  let isTracking = connectAnalytics(browserHistory);

  settings.onChange((key: string) => {
    if (key !== SettingKeys.ACCEPTS_TRACKING) {
      return;
    }

    const { acceptsTracking } = settings;
    if (isTracking !== acceptsTracking) {
      isTracking = connectAnalytics(browserHistory);
    }
  });

  return browserHistory;
};

export default setupAnalytics;
