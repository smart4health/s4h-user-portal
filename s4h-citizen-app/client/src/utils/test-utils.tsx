import { ThemeProvider as MuiThemeProvider } from '@material-ui/core/styles';
import { I18nextProvider } from 'react-i18next';
import { render } from '@testing-library/react';
import React, { ReactElement } from 'react';
import { configureStore } from '@reduxjs/toolkit';
import { Provider as ReduxProvider } from 'react-redux';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';
// @ts-ignore
import { FeatureToggleProvider } from 'react-feature-toggles';
import rootReducer from '../redux/reducers';
import GHTheme from '../Material';
import { featureNames, FeatureNamesValues, Flags } from '../config/flags';
import i18n from '../mocks/i18n';
import { customizeInitialState } from '../store/state';
import { RootState } from '../types';
import { Provider as ReactWaterfallProvider } from '../store';

const featureFlagsAllEnabled: Flags = (
  Object.values(featureNames) as FeatureNamesValues[]
).reduce((accumulator, feature) => {
  accumulator[feature] = true;
  return accumulator;
}, {} as Flags);

/* Custom render function which helps us to have a configurable redux store with defined initial state*/

const history = createMemoryHistory();

const customRender = (
  ui: ReactElement,
  {
    // Ignoring below as a way to type this rightly to Partial<StoreState> was not found
    // @ts-ignore
    initialState,
    store = configureStore({
      reducer: rootReducer,
      preloadedState: initialState,
      middleware: getDefaultMiddleware => getDefaultMiddleware(),
    }),
    extendedWaterfallState = {} as Partial<RootState>,
    ...renderOptions
  } = {}
) => {
  const AllTheProviders: React.FC = props => {
    customizeInitialState({
      flags: featureFlagsAllEnabled,
      ...extendedWaterfallState,
    });
    return (
      <Router history={history}>
        <FeatureToggleProvider featureToggleList={featureFlagsAllEnabled}>
          <ReduxProvider store={store}>
            <ReactWaterfallProvider>
              <I18nextProvider i18n={i18n}>
                <MuiThemeProvider theme={GHTheme}>{props.children}</MuiThemeProvider>
              </I18nextProvider>
            </ReactWaterfallProvider>
          </ReduxProvider>
        </FeatureToggleProvider>
      </Router>
    );
  };

  return render(ui, { wrapper: AllTheProviders, ...renderOptions });
};

export * from '@testing-library/react';
export { customRender as render, history };
