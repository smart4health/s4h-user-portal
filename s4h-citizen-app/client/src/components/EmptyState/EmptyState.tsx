import classNames from 'classnames';
import React from 'react';
import './EmptyState.scss';

type Props = {
  header?: React.ReactNode;
  headerClass?: string;
  content?: React.ReactNode;
  contentClass?: string;
  footer?: React.ReactNode;
  footerClass?: string;
  className?: string;
};

export const emptyStateButtonClasses = 'button--block button--uppercase';

const EmptyState = (props: Props) => {
  const {
    header,
    headerClass,
    content,
    contentClass,
    footer,
    footerClass,
    className,
  } = props;
  return (
    <div
      className={classNames('EmptyState', className)}
      data-test="emptyStateContainer"
    >
      {header && (
        <div className={classNames('EmptyState__header', headerClass)}>{header}</div>
      )}
      {content && (
        <div className={classNames('EmptyState__content', contentClass)}>
          {content}
        </div>
      )}
      {footer && (
        <div className={classNames('EmptyState__footer', footerClass)}>{footer}</div>
      )}
    </div>
  );
};

export default EmptyState;
