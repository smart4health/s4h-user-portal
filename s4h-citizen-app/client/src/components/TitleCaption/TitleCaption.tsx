import React from 'react';
import './TitleCaption.scss';

type Props = {
  title?: string;
  subtitle?: string;
};

const TitleCaption = (props: Props) => {
  const { title, subtitle } = props;
  return (
    <>
      {title && <div className="TitleCaption__title">{title}</div>}
      {subtitle && <div className="TitleCaption__subtitle">{subtitle}</div>}
    </>
  );
};

TitleCaption.defaultProps = {
  subtitle: null,
};

export default TitleCaption;
