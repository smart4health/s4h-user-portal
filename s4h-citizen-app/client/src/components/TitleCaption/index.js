import TitleCaption from './TitleCaption';

export default TitleCaption;
