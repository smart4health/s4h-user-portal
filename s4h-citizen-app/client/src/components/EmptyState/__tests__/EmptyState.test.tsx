import React from 'react';
import EmptyState from '..';
import { render } from '../../../utils/test-utils';
import TitleCaption from '../../TitleCaption';

describe('EmptyState component', () => {
  it('renders without crashing', () => {
    const { container } = render(
      <EmptyState
        header={<TitleCaption title="Title" subtitle="Subtitle" />}
        headerClass="header-class"
        content={<div>Content</div>}
        contentClass="content-class"
        footer={<div>Footer</div>}
        footerClass="footer-class"
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
