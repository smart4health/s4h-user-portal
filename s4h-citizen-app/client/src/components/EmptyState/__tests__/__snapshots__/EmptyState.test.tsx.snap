// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmptyState component renders without crashing 1`] = `
<div
  class="EmptyState"
  data-test="emptyStateContainer"
>
  <div
    class="EmptyState__header header-class"
  >
    <div
      class="TitleCaption__title"
    >
      Title
    </div>
    <div
      class="TitleCaption__subtitle"
    >
      Subtitle
    </div>
  </div>
  <div
    class="EmptyState__content content-class"
  >
    <div>
      Content
    </div>
  </div>
  <div
    class="EmptyState__footer footer-class"
  >
    <div>
      Footer
    </div>
  </div>
</div>
`;
